test/unit/contract_spec.ts

```diff
@@ -0,0 +1,277 @@
+    }
+    return fn.outputs()[0];
+  }
+
+  function roundtrip(funcName: string, input: any, typeName?: string) {
+    let type = getResultType(funcName);
+    let ty = typeName ?? funcName;
+    let obj: any = {};
+    obj[ty] = input;
+    let scVal = SPEC.funcArgsToScVals(funcName, obj)[0];
+    let result = SPEC.scValToNative(scVal, type);
+    expect(result).deep.equal(input);
+  }
+
+  it("u32", () => {
+    roundtrip("u32_", 1);
+  });
+
+  it("i32", () => {
+    roundtrip("i32_", -1);
+  });
+
+  it("i64", () => {
+    roundtrip("i64_", 1n);
+  });
+
+  it("strukt", () => {
+    roundtrip("strukt", { a: 0, b: true, c: "hello" });
+  });
+
+  describe("simple", () => {
+    it("first", () => {
+      const simple = { tag: "First", values: undefined } as const;
+      roundtrip("simple", simple);
+    });
+    it("simple second", () => {
+      const simple = { tag: "Second", values: undefined } as const;
+      roundtrip("simple", simple);
+    });
+
+    it("simple third", () => {
+      const simple = { tag: "Third", values: undefined } as const;
+      roundtrip("simple", simple);
+    });
+  });
+
+  describe("complex", () => {
+    it("struct", () => {
+      const complex = {
+        tag: "Struct",
+        values: [{ a: 0, b: true, c: "hello" }],
+      } as const;
+      roundtrip("complex", complex);
+    });
+
+    it("tuple", () => {
+      const complex = {
+        tag: "Tuple",
+        values: [
+          [
+            { a: 0, b: true, c: "hello" },
+            { tag: "First", values: undefined },
+          ],
+        ],
+      } as const;
+      roundtrip("complex", complex);
+    });
+
+    it("enum", () => {
+      const complex = {
+        tag: "Enum",
+        values: [{ tag: "First", values: undefined }],
+      } as const;
+      roundtrip("complex", complex);
+    });
+
+    it("asset", () => {
+      const complex = { tag: "Asset", values: [addr, 1n] } as const;
+      roundtrip("complex", complex);
+    });
+
+    it("void", () => {
+      const complex = { tag: "Void", values: undefined } as const;
+      roundtrip("complex", complex);
+    });
+  });
+
+  it("addresse", () => {
+    roundtrip("addresse", addr);
+  });
+
+  it("bytes", () => {
+    const bytes = Buffer.from("hello");
+    roundtrip("bytes", bytes);
+  });
+
+  it("bytes_n", () => {
+    const bytes_n = Buffer.from("123456789"); // what's the correct way to construct bytes_n?
+    roundtrip("bytes_n", bytes_n);
+  });
+
+  it("card", () => {
+    const card = 11;
+    roundtrip("card", card);
+  });
+
+  it("boolean", () => {
+    roundtrip("boolean", true);
+  });
+
+  it("not", () => {
+    roundtrip("boolean", false);
+  });
+
+  it("i128", () => {
+    roundtrip("i128", -1n);
+  });
+
+  it("u128", () => {
+    roundtrip("u128", 1n);
+  });
+
+  it("map", () => {
+    const map = new Map();
+    map.set(1, true);
+    map.set(2, false);
+    // map.set(3, 'hahaha') // should throw an error
```

Comment:
can we test this?
```suggestion
    expect(() => map.set(3, 'hahaha')).to.throw(/what error?/i);
```

Reply:
JS doesn't care if the values have conflicting types. So I added a test that tests the error you get when trying to do the round trip again after adding the string.